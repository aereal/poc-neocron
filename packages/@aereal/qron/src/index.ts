export * from "./transactional-task";
export * from "./run-transaction";
