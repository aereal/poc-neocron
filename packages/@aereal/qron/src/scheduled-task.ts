import { Construct } from "@aws-cdk/core";
import { ITable, AttributeType } from "@aws-cdk/aws-dynamodb";
import {
  StateMachine,
  Task,
  Choice,
  Condition,
  IChainable,
  Fail,
  Succeed,
} from "@aws-cdk/aws-stepfunctions";
import { Rule } from "@aws-cdk/aws-events";
import { SfnStateMachine } from "@aws-cdk/aws-events-targets";
import { AttributeValue, UpdateItemTask } from "./dynamodb-sfn-task";

export interface ScheduledTaskProps {
  readonly lockTable: ITable;
  readonly invokeMain: Task;
  readonly taskName: string;
  readonly invocationRule: Rule;
}

const keyTaskName = "taskName";

export class ScheduledTask extends Construct {
  constructor(scope: Construct, id: string, props: ScheduledTaskProps) {
    super(scope, id);

    const { lockTable, invokeMain, taskName, invocationRule } = props;

    const lockKey: AttributeValue = {
      name: keyTaskName,
      type: AttributeType.STRING,
      value: taskName,
    };

    const getLockResultPath = "$.Lock";

    const amount: AttributeValue = {
      name: ":amount",
      type: AttributeType.NUMBER,
      value: "1",
    };

    const getLock = new Task(this, "GetLock", {
      task: new UpdateItemTask({
        table: lockTable,
        parameters: {
          key: lockKey,
          updateExpression:
            "SET handledCount = if_not_exists(handledCount, :default) + :amount",
          expressionAttributeValues: [
            {
              name: ":default",
              type: AttributeType.NUMBER,
              value: "0",
            },
            amount,
          ],
          returnValues: "ALL_NEW",
        },
      }),
      resultPath: getLockResultPath,
    });

    const freeLock = (id: string): Task =>
      new Task(this, id, {
        task: new UpdateItemTask({
          table: lockTable,
          parameters: {
            key: lockKey,
            updateExpression: "SET handledCount = handledCount - :amount",
            expressionAttributeValues: [amount],
            returnValues: "ALL_NEW",
          },
        }),
      });

    // TODO: otherwise
    // TODO: retry?
    const checkLock = (next: IChainable): Choice =>
      new Choice(this, "CheckLock")
        .when(
          Condition.stringEquals(
            `${getLockResultPath}.Attributes.handledCount.N`,
            "1"
          ),
          next
        )
        .otherwise(freeLock("FailedLockFreed").next(new Fail(this, "Finite")));

    const stateMachine = new StateMachine(this, "StateMachine", {
      definition: getLock.next(
        checkLock(
          invokeMain.next(
            freeLock("SuccessFreeLock").next(new Succeed(this, "Succeed"))
          )
        )
      ),
    });

    invocationRule.addTarget(new SfnStateMachine(stateMachine));
  }
}
