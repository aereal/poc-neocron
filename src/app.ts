import { App, AppProps, Environment } from "@aws-cdk/core";

interface NeocronAppProps extends AppProps {
  readonly env: Environment;
}

export class NeocronApp extends App {
  static newFromContext = (): NeocronApp => {
    const { CDK_DEFAULT_ACCOUNT, CDK_DEFAULT_REGION } = process.env;
    if (CDK_DEFAULT_ACCOUNT === undefined) {
      throw new Error("default account not found");
    }
    if (CDK_DEFAULT_REGION === undefined) {
      throw new Error("default region not found");
    }
    return new NeocronApp({
      env: { account: CDK_DEFAULT_ACCOUNT, region: CDK_DEFAULT_REGION },
    });
  };

  private constructor(props: NeocronAppProps) {
    super(props);
  }
}
