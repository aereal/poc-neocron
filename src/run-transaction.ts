import { ExpressionBuilder } from "@aereal/cdk-dynamodb-expression";
import { ITable } from "@aws-cdk/aws-dynamodb";
import {
  Choice,
  Condition,
  Fail,
  INextable,
  State,
  StateMachineFragment,
  TaskStateBase,
} from "@aws-cdk/aws-stepfunctions";
import {
  DynamoAttributeValue,
  DynamoReturnValues,
  DynamoUpdateItem,
} from "@aws-cdk/aws-stepfunctions-tasks";
import { Construct } from "@aws-cdk/core";

export interface RunTransactionalTaskProps {
  /**
   * DynamoDB table manages tasks concurrency.
   * It must have partition key that's name is `taskName` and type is string.
   */
  readonly lockTable: ITable;

  /**
   * Main task state.
   */
  readonly invokeMain: TaskStateBase;

  /**
   * taskName is used as lock key so it must be unique in same [[lockTable]].
   */
  readonly taskName: string;
}

const keyTaskName = "taskName";

/**
 * RunTransaction executes given main task and manages its concurrency.
 *
 * Most common AWS services guarantee at-least-once execution but not exactly-once.
 *
 * The RunTransaction uses Step Function as task runner and DynamoDB as exclusive lock.
 *
 * So we can run some task that supported by Step Functions task in exactly-once manner if max concurrency is 1.
 */
export class RunTransactionalTask extends StateMachineFragment {
  public readonly startState: State;
  public readonly endStates: INextable[];
  private readonly getLockResultPath: string = "$.Lock";
  private readonly lockTable: ITable;
  private readonly taskName: string;
  private readonly amount: DynamoAttributeValue = DynamoAttributeValue.fromNumber(
    1
  );
  private readonly getLockExprBuilder: ExpressionBuilder;
  private readonly freeLockExprBuilder: ExpressionBuilder;

  constructor(scope: Construct, id: string, props: RunTransactionalTaskProps) {
    super(scope, id);

    const { lockTable, invokeMain, taskName } = props;
    this.lockTable = lockTable;
    this.taskName = taskName;
    this.getLockExprBuilder = new ExpressionBuilder();
    this.freeLockExprBuilder = new ExpressionBuilder();

    const getLockResultPath = "$.Lock";
    const assumeLockFreed = this.freeLockTask("AssumeLockFreed");
    const onSuccess = this.freeLockTask("SuccessFreeLock");
    const failedLockFreed = this.freeLockTask("FailedLockFreed");
    const invokeWithRecover = invokeMain
      .addCatch(assumeLockFreed)
      .next(onSuccess);
    const checkLock = new Choice(this, "CheckLock")
      .when(
        Condition.stringEquals(
          `${getLockResultPath}.Attributes.handledCount.N`,
          "1"
        ),
        invokeWithRecover
      )
      .otherwise(failedLockFreed.next(new Fail(this, "Finite")))
      .afterwards();
    const getLock = this.getLockTask();
    const definition = getLock.next(checkLock);
    this.startState = definition.startState;
    this.endStates = [...onSuccess.endStates, ...assumeLockFreed.endStates];
  }

  /**
   *
   */
  private getLockTask(): DynamoUpdateItem {
    const defaultCount = DynamoAttributeValue.fromNumber(0);
    const {
      expression,
      expressionAttributeNames,
      expressionAttributeValues,
    } = this.getLockExprBuilder
      .expr`SET handledCount = if_not_exists(handledCount, ${defaultCount}) + ${this.amount}`;
    return new DynamoUpdateItem(this, "GetLock", {
      table: this.lockTable,
      key: this.lockKey,
      updateExpression: expression,
      expressionAttributeNames,
      expressionAttributeValues,
      returnValues: DynamoReturnValues.ALL_NEW,
      resultPath: this.getLockResultPath,
    });
  }

  /**
   *
   */
  private freeLockTask(id: string): DynamoUpdateItem {
    const {
      expression,
      expressionAttributeValues,
      expressionAttributeNames,
    } = this.freeLockExprBuilder
      .expr`SET handledCount = handledCount - ${DynamoAttributeValue.fromNumber(
      1
    )}`;
    return new DynamoUpdateItem(this, id, {
      table: this.lockTable,
      key: this.lockKey,
      updateExpression: expression,
      expressionAttributeNames,
      expressionAttributeValues,
      returnValues: DynamoReturnValues.ALL_NEW,
    });
  }

  /**
   *
   */
  private get lockKey(): { [key: string]: DynamoAttributeValue } {
    return { [keyTaskName]: DynamoAttributeValue.fromString(this.taskName) };
  }
}
