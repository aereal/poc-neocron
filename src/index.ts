import { QronApp } from "./app";

QronApp.newFromContext().synth();
