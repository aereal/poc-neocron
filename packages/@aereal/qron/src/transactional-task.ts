import { ITable } from "@aws-cdk/aws-dynamodb";
import {
  IStateMachine,
  StateMachine,
  Succeed,
  TaskStateBase,
} from "@aws-cdk/aws-stepfunctions";
import { Construct } from "@aws-cdk/core";
import { RunTransaction } from "./run-transaction";

export interface TransactionalTaskProps {
  /**
   * DynamoDB table manages tasks concurrency.
   * It must have partition key that's name is `taskName` and type is string.
   */
  readonly lockTable: ITable;

  readonly invokeMain: TaskStateBase;

  /**
   * taskName is used as lock key so it must be unique in same [[lockTable]].
   */
  readonly taskName: string;
}

/**
 * TransactionalTask executes given main task and manages its concurrency.
 *
 * Most common AWS services guarantee at-least-once execution but not exactly-once.
 *
 * The TransactionalTask uses Step Function as task runner and DynamoDB as exclusive lock.
 *
 * So we can run some task that supported by Step Functions task in exactly-once manner if max concurrency is 1.
 *
 * @deprecated
 */
export class TransactionalTask extends Construct {
  public readonly stateMachine: IStateMachine;

  constructor(scope: Construct, id: string, props: TransactionalTaskProps) {
    super(scope, id);

    this.stateMachine = new StateMachine(this, "StateMachine", {
      definition: new RunTransaction(this, "MainState", props).next(
        new Succeed(this, "Succeed")
      ),
    });
  }
}
