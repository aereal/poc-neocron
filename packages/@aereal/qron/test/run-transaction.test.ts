import { SynthUtils } from "@aws-cdk/assert";
import { AttributeType, Table } from "@aws-cdk/aws-dynamodb";
import { Topic } from "@aws-cdk/aws-sns";
import { StateMachine, TaskInput } from "@aws-cdk/aws-stepfunctions";
import { SnsPublish } from "@aws-cdk/aws-stepfunctions-tasks";
import { Stack } from "@aws-cdk/core";
import { RunTransaction } from "../src/run-transaction";

describe("TransactionalTask", () => {
  test("ok", () => {
    const stack = new Stack();
    const lockTable = new Table(stack, "LockTable", {
      partitionKey: {
        name: "id",
        type: AttributeType.STRING,
      },
    });
    const topic = new Topic(stack, "Topic");
    new StateMachine(stack, "StateMachine", {
      definition: new RunTransaction(stack, "RunTransaction", {
        lockTable,
        taskName: "test-task",
        invokeMain: new SnsPublish(stack, "PublishTopic", {
          topic,
          message: TaskInput.fromObject({}),
        }),
      }),
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
