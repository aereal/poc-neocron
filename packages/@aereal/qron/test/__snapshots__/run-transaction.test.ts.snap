// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionalTask ok 1`] = `
Object {
  "Resources": Object {
    "LockTableB9DACF42": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "StateMachine2E01A3A5": Object {
      "DependsOn": Array [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"GetLock\\",\\"States\\":{\\"GetLock\\":{\\"Next\\":\\"CheckLock\\",\\"Type\\":\\"Task\\",\\"ResultPath\\":\\"$.Lock\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::dynamodb:updateItem\\",\\"Parameters\\":{\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"ExpressionAttributeValues\\":{\\":v3\\":{\\"N\\":\\"0\\"},\\":v4\\":{\\"N\\":\\"1\\"}},\\"ReturnValues\\":\\"ALL_NEW\\",\\"UpdateExpression\\":\\"SET handledCount = if_not_exists(handledCount, :v3) + :v4\\"}},\\"CheckLock\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Lock.Attributes.handledCount.N\\",\\"StringEquals\\":\\"1\\",\\"Next\\":\\"PublishTopic\\"}],\\"Default\\":\\"FailedLockFreed\\"},\\"FailedLockFreed\\":{\\"Next\\":\\"Finite\\",\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::dynamodb:updateItem\\",\\"Parameters\\":{\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"ExpressionAttributeValues\\":{\\":v2\\":{\\"N\\":\\"1\\"}},\\"ReturnValues\\":\\"ALL_NEW\\",\\"UpdateExpression\\":\\"SET handledCount = handledCount - :v2\\"}},\\"Finite\\":{\\"Type\\":\\"Fail\\"},\\"PublishTopic\\":{\\"Next\\":\\"SuccessFreeLock\\",\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"AssumeLockFreed\\"}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish\\",\\"Parameters\\":{\\"TopicArn\\":\\"",
              Object {
                "Ref": "TopicBFC7AF6E",
              },
              "\\",\\"Message\\":{}}},\\"SuccessFreeLock\\":{\\"End\\":true,\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::dynamodb:updateItem\\",\\"Parameters\\":{\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"ExpressionAttributeValues\\":{\\":v1\\":{\\"N\\":\\"1\\"}},\\"ReturnValues\\":\\"ALL_NEW\\",\\"UpdateExpression\\":\\"SET handledCount = handledCount - :v1\\"}},\\"AssumeLockFreed\\":{\\"End\\":true,\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::dynamodb:updateItem\\",\\"Parameters\\":{\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"ExpressionAttributeValues\\":{\\":v0\\":{\\"N\\":\\"1\\"}},\\"ReturnValues\\":\\"ALL_NEW\\",\\"UpdateExpression\\":\\"SET handledCount = handledCount - :v0\\"}}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "StateMachineRoleB840431D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:UpdateItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/",
                    Object {
                      "Ref": "LockTableB9DACF42",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TopicBFC7AF6E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": Array [
          Object {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TopicBFC7AF6E": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
